nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,354 @@
+    step, in wchich the alphas are frozen and the model parameters are optimized. In this step, the
+    out put of each choice are sumed up with respect to their alpha value. The result of DARTS
+    is argmax in alpha.
+    The DARTS Model should be trained with ParallelTraiValDataloader in nn.retiarii.oneshot.pytorch.utils.
+    See base class for more attributes.
+
+    Reference
+    ----------
+    [1]H. Liu, K. Simonyan, and Y. Yang, “DARTS: Differentiable Architecture Search,” presented at the
+    International Conference on Learning Representations, Sep. 2018. Available: https://openreview.net/forum?id=S1eYHoC5FX
+    '''
+    def __init__(self, model, custom_replace_dict = None):
+        '''
+        Parameters
+        ----------
+        base_model : pl.LightningModule
+            The module in evaluators in nni.retiarii.evaluator.lightning. User defined model
+            is wrapped by base_model, and base_model will be wrapped by this model.
+        custom_replace_dict : dict{ type : callable }, default = None
+            The custom xxxChoice replace method. Keys should be xxxChoice type and values should 
+            return an nn.module. This custom replace dict will override the default replace
+            dict of each nas method.
+        '''
+        super().__init__(model, custom_replace_dict)
+        self.automatic_optimization = False
+    
+    def on_train_start(self) -> None:
+        return super().on_train_start()
+        
+    def training_step(self, batch, batch_idx):
+        # grad manually, only 1 architecture optimizer for darts
+        opts = self.optimizers()
+        arc_optim = opts[0]
+        w_optim = opts[1:]
+
+        # ParallelTrainValDataLoader will yield both train and val data in a batch
+        trn_batch, val_batch = batch
+
+        # phase 1. architecture step
+        # The resample_architecture hook is kept for some following darts-based nas
+        # methods such as proxyless. See code of those methods for details. 
+        self.resample_architecture()
+        arc_optim.zero_grad()
+        arc_step_loss = self.model.training_step(val_batch, 2 * batch_idx)
+        self.manual_backward(arc_step_loss)
+        self.finalize_grad()
+        arc_optim.step()
+
+        # phase 2: model step
+        self.resample_architecture()
+        for opt in w_optim:
+            opt.zero_grad()
+        w_step_loss = self.model.training_step(trn_batch, 2 * batch_idx + 1)
+        self.manual_backward(w_step_loss)
+        for opt in w_optim:
+            opt.step()
+        return w_step_loss
+
+    def resample_architecture(self):
+        '''
+        Hook kept for darts-based methods. Some following works resample the architecture to
+        reduce the memory consumption during training to fit the method to bigger models. Details
+        are provided in code of those algorithms.
+        '''
+        pass
+
+    def finalize_grad(self):
+        '''
```

Comment:
This is an implementation note rather than docstring that should be visible to anyone who uses this.
So it should be `# NOTE: this hook is kept currently for ProxylessNAS`.